frontend/src/main/scala/bloop/dap/DebugSession.scala

```diff
@@ -0,0 +1,182 @@
+  private type LaunchId = Int
+
+  // A set of all processed launched requests by the client
+  private val launchedRequests = mutable.Set.empty[LaunchId]
+
+  private val debuggeeExited = Promise[Unit]()
+  private val debugAddress = Promise[InetSocketAddress]()
+  private val exitStatusPromise = Promise[DebugSession.ExitStatus]()
+
+  private val isStarted = Atomic(false)
+  private val isCancelled = Atomic(false)
+
+  // Access to this cancelable is always protected behind `isCancelled` to avoid race conditions
+  private val runningDebuggee = Atomic(CancelableFuture.unit)
+
+  /**
+   * Schedules the start of the debugging session.
+   *
+   * For a session to start, two executions must happen independently in a
+   * non-blocking way: the debuggee process is started in the background and
+   * the DAP server starts listening to client requests in an IO thread.
+   */
+  def startDebuggeeAndServer(): Unit = {
+    ioScheduler.executeAsync(() => {
+      if (isStarted.compareAndSet(false, true)) {
+        val logger =
+          new DebugSessionLogger(this, address => debugAddress.success(address), initialLogger)
+        val debuggeeHandle = startDebuggee(logger).runAsync(ioScheduler)
+        isCancelled.getAndTransform { isCancelled =>
+          if (isCancelled) debuggeeHandle.cancel()
+          else runningDebuggee.set(debuggeeHandle)
+          isCancelled
+        }
+
+        try super.run()
+        finally {
+          exitStatusPromise.trySuccess(DebugSession.Terminated); ()
+        }
+      }
+    })
+  }
+
+  override def dispatchRequest(request: Request): Unit = {
+    val requestId = request.seq
+    request.command match {
+      case "launch" =>
+        launchedRequests.add(requestId)
+        val _ = Task
+          .fromFuture(debugAddress.future)
+          .map(DebugSession.toAttachRequest(requestId, _))
+          .foreachL(super.dispatchRequest)
+          .runAsync(ioScheduler)
+
+      case "disconnect" =>
+        if (DebugSession.shouldRestart(request))
+          exitStatusPromise.trySuccess(DebugSession.Restarted)
+        runningDebuggee.get.cancel()
+        super.dispatchRequest(request)
+
+      case _ => super.dispatchRequest(request)
+    }
+  }
+
+  override def sendResponse(response: Response): Unit = {
+    val requestId = response.request_seq
+    response.command match {
+      case "attach" if launchedRequests(requestId) =>
+        // Trick dap4j into thinking we're processing a launch instead of attach
+        response.command = Command.LAUNCH.getName
+        super.sendResponse(response)
+      case "disconnect" if requestId == DebugSession.InternalRequestId =>
+        // Request sent by the session itself, don't return to client
+        socket.close()
+      case _ => super.sendResponse(response)
+    }
+  }
+
+  override def sendEvent(event: Events.DebugEvent): Unit = {
+    if (event.`type` == "exited") {
+      debuggeeExited.success(())
+    }
+
+    super.sendEvent(event)
```

Comment:
I moved this line after the `if` so that if this throws, for any reason, it doesn't affect the completion of `debuggeeExited`. Also, it looks like there might be a race condition here and we might never be completing `debuggeeExited` because the exited event is not generated if the socket was already closed. I haven't confirmed if this race condition indeed happens, it's just an intuition. What are your thoughts?

Reply:
I am once again completing this promise after the event is being sent. It must be this way, since we are closing the socket once it is completed. If we were completing before sending event, we would get a race (closing socket, sending event).